Add Button component

ashleemboyer opened the pull request:
## "Ready for Review" checklist

<!--
Check these items off in the GitHub PR UI as you complete them.
-->

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## Relevant links

<!-- 
Make sure to remove any '.' characters from the branch name 
Example: ks.my-branch -> ksmy-branch
-->

- [Preview link](https://dev-portal-git-ambadd-button-component-hashicorp.vercel.app/swingset/components/button) 🔎
- [Asana task](https://app.asana.com/0/1201979821989355/1201654639085740/f) 🎟️

## What/Why

<!--
Briefly list out the changes proposed in this PR.
-->

- Adds new `Button` component along with a swingset page for testing
- Adds `copy` icon option for `SwingsetTestIcon` so `Button` tests are more interesting

## How

<!--
Dive into the approach you took, list resources you referenced, detail other approaches you tried but didn't end up going with, etc.
-->

Resources used:

- [HDS Figma designs](https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/HDS-Product---Components-%5BWIP%5D?node-id=2340%3A21001)
- [HDS Button previews](https://design-system-components-hashicorp.vercel.app/components/button)

## Testing

<!--
Create a checklist for going through how to test your proposed changes. If there is anything to configure before interacting with the project in a browser, such as toggling feature flags, changing machine settings, or simulating behavior in browser dev tools, list those steps first.

- [ ] Step 1
- [ ] Step 2
- [ ] Step 3
- [ ] ...
-->

Explore [the Button swingset page](https://dev-portal-git-ambadd-button-component-hashicorp.vercel.app/swingset/components/button) and see if everything behaves as you expect.

## Anything else?

<!--
If there is anything you came across that you chose not to address in this PR but plan to soon, list those items here and any Asana tasks you created to go with them.
-->

Nope!

ashleemboyer commented:
FYI I still need to fill in the props table on the swingset page, but this is ready for review otherwise. 😊

zchsh commented on src/components/button/index.tsx:
```diff
@@ -0,0 +1,64 @@
+    )
+  }
+
+  if (hasIcon && !hasLabel) {
```
Noticed the Swingset examples for `Icon and Text` are throwing this error. If we have text in the button, then I think we don't need the label, right?

```suggestion
  if (hasIcon && !hasLabel && !hasText) {
```

ashleemboyer replied:
Uh oh! I didn't notice the error message pop up below these, will work on fixing it 😊

ashleemboyer replied:
Fixed in https://github.com/hashicorp/dev-portal/pull/215/commits/c9818e981f0c32e705dceaeb7eb6a142702fd844! Added an `isIconOnly` boolean and use it in this check now. Great catch! ✨

zchsh commented on src/layouts/core-dev-dot-layout/core-dev-dot-layout.module.css:
```diff
@@ -21,8 +21,12 @@
     font-family: var(--font-body);
 
-    &:hover {
+    &:not(:disabled) {
```
TLDR; very non-blocking random thought, feel free to ignore!

Re: the TODO comment above... would we be open to starting with these styles within the button component?

I'm definitely on the fence here; I can't really imagine a `button` where we wouldn't want these styles... however I think it shouldn't be too much of a tradeoff to repeat these styles in every button-ish thing we build.

Alternately... maybe we split this out to a kind of "normalize"-ish stylesheet? I suspect this might be something the design systems team has an opinion on 🤔

zchsh replied:
I kinda like the idea of "normalize-on-the-fly" - eg if there's some other `<button>`-using components that need to be normalized as well (eg the ones in the sidebar), we'd repeat these styles within that component (and helpers seem like a great way to abstract rather than repeat! 👍).

If we go the more traditional normalize route, I'm really not sure where that stylesheet would live 🤔 i guess in most projects it'd be a global stylesheet; in our case since we have the co-located dot-ios I suppose we'd need something scoped to `core-dev-dot-layout` 🤷‍♂️

ashleemboyer replied:
I really like the idea of using a helper for this! For this cursor thing specifically, we could do something like this in `src/pages/style.css`:

```
.g-button-cursor-pointer {
  &:not(:disabled) {
    cursor: pointer;
  }

  &:disabled {
    cursor: not-allowed;
  }
}
```

I assume we'd want a separate one for handling `font-family`, or maybe that one actually stays here since font probably won't change between buttons? 🤔

zchsh replied:
☝️ yeah i dig that! another approach could be to tie it to the HTML element to make it a bit more generic, and to signal the intent (i think?) to apply this to all `button` elements, eg:

```
/* src/styles/normalize.module.css */
.normalize-button {
  font-family: var(--font-body);
  font-size: 100%;
  line-height: inherit;
  
  &:not(:disabled) {
    cursor: pointer;
  }
 
  &:disabled {
    cursor: not-allowed;
  }
}

/* src/components/some-button/some-button.module.css */
.root {
  composes: normalize-button from "src/styles/normalize.module.css";
  /* ... other implementation-specific styles */
}
```

personally kind of on the fence... a "standard" global `normalize.css` might be better than the above in some ways (though arguably a little less "encapsulated").

in any case, global styles definitely seem more than fine for now 👍 especially since this stuff for `button` feels really genuinely global (rather than being convenient short-term but hard to maintain long-term, which I think can sometimes happen when abstracting to global styles too early).

zchsh reviewed (approved):
✅ Noticed an issue with icon + text button examples in Swingset, but however that gets resolved this looks good to me ☺️

kendallstrautman reviewed (approved):
LGTM! 👏

kendallstrautman commented on src/components/button/types.ts:
```diff
@@ -0,0 +1,19 @@
+type ButtonElementProps = JSX.IntrinsicElements['button']
+
+// TODO: add jsdoc to each prop
+export type ButtonProps = Pick<
```
I think this should be an `interface`! Especially since its extending the button element props.

kendallstrautman replied:
Interesting notes in[ the comments](https://www.typescriptlang.org/play#example/types-vs-interfaces) here re: interfaces v. types

kendallstrautman replied:
It's up to you though 👐

ashleemboyer replied:
Just pushed up some commits that have the outcome of making this an `interface`. 😊

kendallstrautman commented on src/components/button/types.ts:
```diff
@@ -0,0 +1,19 @@
+type ButtonElementProps = JSX.IntrinsicElements['button']
+
+// TODO: add jsdoc to each prop
+export type ButtonProps = Pick<
+  ButtonElementProps,
+  'disabled' | 'onClick' | 'type'
+> & {
```
```suggestion
export interface ButtonProps extends Pick<
  ButtonElementProps,
  'disabled' | 'onClick' | 'type'
> {
```

BRKalow reviewed (approved):
👏

ashleemboyer commented:
Have now added the `id` and `name` props (since those are probably very commonly used, easy to get them out of the way now), and added details to the `types.ts` file and the swingset page.
